### ISSUE TITLE
[RFC] Views with primary keys; indexes

### ISSUE THREAD

**Issue by ryzhyk** [feldera/feldera#2681]
## Motivation

There are several use cases that require views to have primary keys:
* Combine deletes and inserts to the same key into a single atomic update
  * Many sinks either don't support deletions at all or don't support them efficiently, but they do support upserts. Having a primary key on a view will enable us to combine matching deletes and inserts into a single upsert operation
  * Even when the sink supports deletions, we may want to apply the deletion and the insertion atomically to avoid the user-visible intermediate state after the old value was deleted and the new one was inserted
  * For workloads where most changes are upserts, splitting them into two separate operations doubles the amount of output traffic

* Index materialized tables and views for fast lookups
  * In many use cases, users want to query the materialized view and tables as key-value stores.  We can do this via the ad hoc query layer, but it's currently not efficient as every lookup requires a full scan of the collection. Datafusion supports index lookups indirectly via filter pushdown.
  * Note: this does't strictly require a primary key or even a unique key, as there can be multiple values per key. So perhaps this could be a separate mechanism, e.g., `CREATE INDEX`.

## Semantics

We don't have a way to enforce uniqueness of primary keys. The user can write a query that can return multiple values for the same key. We need to decide how to handle such situations.  One option is to signal an error and use some kind of disambiguation policy, e.g., return the largest value associate with the key (this is equivalent to rewriting the view like this `SELECT MAX(*) FROM my_view GROUP BY key`)

## Syntax

`CREATE VIEW my_view WITH ('key_fields' = ['id', 'other_key_field']) AS SELECT ..`

**Comment by mihaibudiu** [comment]
No way this syntax is acceptable.
A better approach is to use a CREATE UNIQUE INDEX statement; these normally don't work on views, but I don't see why they wouldn't.
E.g.: https://learn.microsoft.com/en-us/sql/t-sql/statements/create-index-transact-sql?view=sql-server-ver16

**Comment by ryzhyk** [comment]
> No way this syntax is acceptable. 

Why not? This is similar to for example 'materialize' = 'true' on tables.

> A better approach is to use a CREATE UNIQUE INDEX statement; these normally don't work on views, but I don't see why they wouldn't. E.g.: https://learn.microsoft.com/en-us/sql/t-sql/statements/create-index-transact-sql?view=sql-server-ver16

I agree, this is nicer because it's closer to standard SQL. But there can be many unique (or non-unique for that matter) indexes. So does this mean that output connectors are attached to an index, not the view?

**Comment by mihaibudiu** [comment]
no, I think the index is a property of the view. A unique index is exactly like an UPSERT operator.

**Comment by ryzhyk** [comment]
so, we don't allow multiple indexes?

**Comment by mihaibudiu** [comment]
Maybe we can't create indexes today, but we can create unique indexes.
SQL supports multiple unique indexes on the same table.

**Comment by ryzhyk** [comment]
We can create both. But back to my original question: if there are multiple unique indexes for a view, which one should we use when combining inserts and deletes into upserts and sending them to a connector?

**Comment by mihaibudiu** [comment]
SQL uses all of them.

**Comment by ryzhyk** [comment]
> SQL uses all of them.

this doesn't answer my question

**Comment by mihaibudiu** [comment]
you can apply all of them in any order and you should get the same result I think
